workplace/models.py

```diff
@@ -256,3 +272,35 @@ class Reservation(SafeDeleteModel):
 
     def __str__(self):
         return str(self.user)
+
+    def update_tomatoes(self):
+        """
+        Credit tomatoes to user if he is present or remove them if updated
+        back to not present
+        """
+        from tomato.models import Tomato
+        if self.is_present:
+            # user presence set to True: Add tomato for the timeslot
+            Tomato.objects.create(
+                user=self.user,
+                number_of_tomato=self.timeslot.number_of_tomatoes,
+                source=Tomato.TOMATO_SOURCE_TIMESLOT,
+                content_object=self,
+            )
+        else:
+            # User presence set to False: remove tomato for timeslot if exists
+            reservation_type = ContentType.objects.get_for_model(
+                Reservation)
+            try:
+                Tomato.objects.get(
+                    user=self.user,
+                    source=Tomato.TOMATO_SOURCE_TIMESLOT,
+                    content_type=reservation_type,
+                    object_id=self.id,
+                ).delete()
+            except Tomato.DoesNotExist:
+                pass
+
+    def save(self, *args, **kwargs):
+        super(Reservation, self).save(*args, **kwargs)
+        self.update_tomatoes()
```

Comment:
Im pretty sure that if you update a reservation without changing a `is_present` already `True` you will create a second `Tomato` object!

I think you need some quick enhancements:

1. Make sure to have a check before create to not duplicate (line: ~283)
2. Maybe adding a little condition in the `save()` to avoid calling DB and logic EVERYTIME (ie: via a IF or the usage of Django's `signals`

Reply:
Updated

Reply:
Also I updated the task for retreat tomato to also use get or create